src/Evolution/DiscontinuousGalerkin/Actions/ApplyBoundaryCorrections.hpp

```diff
@@ -0,0 +1,417 @@
+ * correction on each mortar and then lifts it into the volume.
+ *
+ * Future additions include:
+ * - boundary conditions, both through ghost cells and by changing the time
+ *   derivatives.
+ * - support local time stepping (shouldn't be very difficult)
+ */
+template <typename Metavariables>
+struct ApplyBoundaryCorrections {
+  using inbox_tags =
+      tmpl::list<evolution::dg::Tags::BoundaryCorrectionAndGhostCellsInbox<
+          Metavariables::volume_dim>>;
+  using const_global_cache_tags = tmpl::list<
+      evolution::Tags::BoundaryCorrection<typename Metavariables::system>,
+      ::dg::Tags::Formulation>;
+
+  template <typename DbTagsList, typename... InboxTags, typename ArrayIndex,
+            typename ActionList, typename ParallelComponent>
+  static std::tuple<db::DataBox<DbTagsList>&&> apply(
+      db::DataBox<DbTagsList>& box, tuples::TaggedTuple<InboxTags...>& inboxes,
+      const Parallel::GlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& /*array_index*/, ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    constexpr size_t volume_dim = Metavariables::system::volume_dim;
+
+    if (UNLIKELY(db::get<domain::Tags::Element<volume_dim>>(box)
+                     .number_of_neighbors() == 0)) {
+      // We have no neighbors, yay!
+      return {std::move(box)};
+    }
+
+    if constexpr (Metavariables::local_time_stepping) {
+      ERROR("Local time stepping not supported");
+    } else {
+      apply_global_time_stepping(make_not_null(&box), make_not_null(&inboxes));
+    }
+    return {std::move(box)};
+  }
+
+  template <typename DbTags, typename... InboxTags, typename ArrayIndex>
+  static bool is_ready(const db::DataBox<DbTags>& box,
+                       const tuples::TaggedTuple<InboxTags...>& inboxes,
+                       const Parallel::GlobalCache<Metavariables>& /*cache*/,
+                       const ArrayIndex& /*array_index*/) noexcept;
+
+ private:
+  template <typename DbTagsList, typename... InboxTags>
+  static void apply_global_time_stepping(
+      gsl::not_null<db::DataBox<DbTagsList>*> box,
+      gsl::not_null<tuples::TaggedTuple<InboxTags...>*> inboxes) noexcept;
+};
+
+template <typename Metavariables>
+template <typename DbTagsList, typename... InboxTags>
+void ApplyBoundaryCorrections<Metavariables>::apply_global_time_stepping(
+    const gsl::not_null<db::DataBox<DbTagsList>*> box,
+    const gsl::not_null<tuples::TaggedTuple<InboxTags...>*> inboxes) noexcept {
+  constexpr size_t volume_dim = Metavariables::system::volume_dim;
+
+  using Key = std::pair<Direction<volume_dim>, ElementId<volume_dim>>;
+  std::map<
+      TimeStepId,
+      FixedHashMap<
+          maximum_number_of_neighbors(volume_dim), Key,
+          std::tuple<Mesh<volume_dim - 1>, std::optional<std::vector<double>>,
+                     std::optional<std::vector<double>>, ::TimeStepId>,
+          boost::hash<Key>>>& inbox =
```

Comment:
I have found that, especially for these long types where I need to access inside many containers, that it's actually easier if I know what the type is rather than having to repeatedly reference a tag